### PR TITLE
[FlexCounterManager] Add ACL_COUNTER FC type

### PR DESCRIPTION
**Pull request by stepanblyschak**
Signed-off-by: Stepan Blyschak <stepanb@nvidia.com>

**What I did**

**Why I did it**

**How I verified it**

**Details if related**

### CODE REVIEW

File: orchagent/aclorch.h [hunk 1]
```diff
@@ -491,6 +478,8 @@ class AclOrch : public Orch, public Observer
 
     acl_capabilities_t m_aclCapabilities;
     acl_action_enum_values_capabilities_t m_aclEnumActionCapabilities;
+    Table m_acl_counter_rule_map;
```
Comment (stepanblyschak) [inline orchagent/aclorch.h:481]: // make it clear that it is a table

File: orchagent/aclorch.cpp [hunk 2]
```diff
@@ -4083,6 +4101,42 @@ sai_status_t AclOrch::deleteDTelWatchListTables()
     return SAI_STATUS_SUCCESS;
 }
 
+void AclOrch::registerFlexCounter(const AclRule& rule)
+{
+    SWSS_LOG_ENTER();
+
+    string ruleIdentifier = generateAclRuleIdentifierInCountersDb(rule);
```
Comment (stepanblyschak) [inline orchagent/aclorch.cpp:4108]: auto

File: orchagent/aclorch.cpp [hunk 3]
```diff
@@ -3321,6 +3362,7 @@ bool AclOrch::updateAclRule(string table_id, string rule_id, bool enableCounter)
         return false;
     }
 
+    deregisterFlexCounter(*rule);
```
Comment (stepanblyschak) [inline orchagent/aclorch.cpp:3365]: move up before disableCounter()

File: tests/dvslib/dvs_acl.py [hunk 4]
```diff
@@ -517,7 +521,6 @@ def verify_acl_rule_set(
         for entry in acl_entries:
             rule = self.asic_db.wait_for_entry("ASIC_STATE:SAI_OBJECT_TYPE_ACL_ENTRY", entry)
             priority = rule.get("SAI_ACL_ENTRY_ATTR_PRIORITY", None)
-            assert priority in priorities
```
Comment (stepanblyschak) [inline tests/dvslib/dvs_acl.py:520]: this shouldn't be removed

File: orchagent/aclorch.cpp [hunk 5]
```diff
@@ -627,7 +639,19 @@ bool AclRule::remove()
...
+bool AclRule::remove()
+{
+    SWSS_LOG_ENTER();
+
+    if (!removeRule())
+    {
+        return false;
+    }
+
+    auto res = removeRanges();
```
Comment (nazariig) [inline orchagent/aclorch.cpp:654]: @stepanblyschak not sure if it's ok to use `auto` here...

File: orchagent/aclorch.cpp [hunk 6]
```diff
@@ -851,13 +853,12 @@ bool AclRule::createCounter()
...
     counter_attrs.push_back(attr);
 
-    attr.id = SAI_ACL_COUNTER_ATTR_ENABLE_BYTE_COUNT;
-    attr.value.booldata = true;
-    counter_attrs.push_back(attr);
-
-    attr.id = SAI_ACL_COUNTER_ATTR_ENABLE_PACKET_COUNT;
-    attr.value.booldata = true;
-    counter_attrs.push_back(attr);
+    for (auto counterAttrPair: aclCounterLookup)
```
Comment (nazariig) [inline orchagent/aclorch.cpp:856]: @stepanblyschak why not `const auto&` ?

File: orchagent/aclorch.cpp [hunk 7]
```diff
@@ -2137,26 +2145,45 @@ bool AclRuleDTelFlowWatchListEntry::create()
 {
     SWSS_LOG_ENTER();
 
-    if (!m_pDTelOrch)
```
Comment (nazariig) [inline orchagent/aclorch.cpp:2140]: @stepanblyschak why this was removed?

File: orchagent/aclorch.cpp [hunk 8]
```diff
@@ -3165,7 +3183,17 @@ bool AclOrch::addAclRule(shared_ptr<AclRule> newRule, string table_id)
         return false;
     }
 
-    return m_AclTables[table_oid].add(newRule);
+    if (!m_AclTables[table_oid].add(newRule))
```
Comment (nazariig) [inline orchagent/aclorch.cpp:3186]: @stepanblyschak do we need to add `if (!newRule)` prior to this?

File: orchagent/aclorch.cpp [hunk 9]
```diff
@@ -4083,6 +4101,42 @@ sai_status_t AclOrch::deleteDTelWatchListTables()
...
 
+void AclOrch::registerFlexCounter(const AclRule& rule)
+{
+    SWSS_LOG_ENTER();
+
+    auto ruleIdentifier = generateAclRuleIdentifierInCountersDb(rule);
+    auto counterOidStr = sai_serialize_object_id(rule.getCounterOid());
+
+    unordered_set<string> serializedCounterStatAttrs;
+    for (auto counterAttrPair: aclCounterLookup)
```
Comment (nazariig) [inline orchagent/aclorch.cpp:4112]: @stepanblyschak why not to use `const auto&` ?

**Review by nazariig — APPROVED** [review]

File: orchagent/aclorch.cpp (hunk 5 above)
Comment (stepanblyschak) [reply orchagent/aclorch.cpp:654]: This is questionable. As a rule of thumb I would prefer the code with auto, especially considering that return type of the function can be also deduced in newer c++.  I don't see a value in having to explicitelly say it's bool.

File: orchagent/aclorch.cpp (hunk 6 above)
Comment (stepanblyschak) [reply orchagent/aclorch.cpp:856]: Better to use.

File: orchagent/aclorch.cpp (hunk 9 above)
Comment (stepanblyschak) [reply orchagent/aclorch.cpp:4112]: Better to use.

File: orchagent/aclorch.cpp (hunk 8 above)
Comment (stepanblyschak) [reply orchagent/aclorch.cpp:3186]: We do, currently it is undefined behaviour if we pass nullptr regardless of this change. I will try to push it in another PR.

**Pull request closed without merging**